server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java

```diff
@@ -128,26 +127,34 @@ public static AggregatedDfs aggregateDfs(Collection<DfsSearchResult> results) {
         return new AggregatedDfs(termStatistics, fieldStatistics, aggMaxDoc);
     }
 
-    public static DfsKnnResults mergeKnnResults(SearchRequest request, List<DfsSearchResult> dfsSearchResults) {
+    public static List<DfsKnnResults> mergeKnnResults(SearchRequest request, List<DfsSearchResult> dfsSearchResults) {
         if (request.hasKnnSearch() == false) {
             return null;
         }
 
-        List<TopDocs> topDocs = new ArrayList<>();
-        for (DfsSearchResult dfsSearchResult : dfsSearchResults) {
-            if (dfsSearchResult.knnResults() != null) {
-                ScoreDoc[] scoreDocs = dfsSearchResult.knnResults().scoreDocs();
-                TotalHits totalHits = new TotalHits(scoreDocs.length, Relation.EQUAL_TO);
+        List<List<TopDocs>> topDocsLists = new ArrayList<>(request.source().knnSearch().size());
+        for (int i = 0; i < request.source().knnSearch().size(); i++) {
+            topDocsLists.add(new ArrayList<>());
+        }
 
-                TopDocs shardTopDocs = new TopDocs(totalHits, scoreDocs);
-                setShardIndex(shardTopDocs, dfsSearchResult.getShardIndex());
-                topDocs.add(shardTopDocs);
+        for (DfsSearchResult dfsSearchResult : dfsSearchResults) {
+            if (dfsSearchResult.knnResults() != null && dfsSearchResult.knnResults().isEmpty() == false) {
```

Comment:
nit: we could skip the isEmpty conditional, if empty the following loop will do nothing.